Cabal/src/Distribution/Simple/Glob.hs

```diff
@@ -283,73 +493,153 @@ matchDirFileGlobWithDie verbosity rip version dir filepath = case parseFileGlob
-      joinedPrefix = joinPath prefixSegments
-  case final of
-    FinalMatch recursive multidot exts -> do
-      let prefix = dir </> joinedPrefix
-      directoryExists <- doesDirectoryExist prefix
-      if directoryExists
-        then do
-          candidates <- case recursive of
-            Recursive -> getDirectoryContentsRecursive prefix
-            NonRecursive -> filterM (doesFileExist . (prefix </>)) =<< getDirectoryContents prefix
-          let checkName candidate = do
-                let (candidateBase, candidateExts) = splitExtensions $ takeFileName candidate
-                guard (not (null candidateBase))
-                match <- checkExt multidot exts candidateExts
-                return (joinedPrefix </> candidate <$ match)
-          return $ mapMaybe checkName candidates
-        else return [GlobMissingDirectory joinedPrefix]
-    FinalLit Recursive fn -> do
-      let prefix = dir </> joinedPrefix
-      directoryExists <- doesDirectoryExist prefix
-      if directoryExists
-        then do
-          candidates <- getDirectoryContentsRecursive prefix
-          let checkName candidate
-                | takeFileName candidate == fn = Just $ GlobMatch (joinedPrefix </> candidate)
-                | otherwise = Nothing
-          return $ mapMaybe checkName candidates
-        else return [GlobMissingDirectory joinedPrefix]
-    FinalLit NonRecursive fn -> do
-      exists <- doesFileExist (dir </> joinedPrefix </> fn)
-      return [GlobMatch (joinedPrefix </> fn) | exists]
-
-unfoldr' :: (a -> Either r (b, a)) -> a -> ([b], r)
-unfoldr' f a = case f a of
-  Left r -> ([], r)
-  Right (b, a') -> case unfoldr' f a' of
-    (bs, r) -> (b : bs, r)
-
--- | Extract the (possibly null) constant prefix from the pattern.
--- This has the property that, if @(pref, final) = splitConstantPrefix pat@,
--- then @pat === foldr GlobStem (GlobFinal final) pref@.
-splitConstantPrefix :: Glob -> ([FilePath], GlobFinal)
-splitConstantPrefix = unfoldr' step
+  let
+    (prefixSegments, variablePattern) = splitConstantPrefix pat
+    joinedPrefix = joinPath prefixSegments
+
+    -- The glob matching function depends on whether we care about the cabal version or not
+    doesGlobMatch :: Glob -> String -> Maybe (GlobResult ())
+    doesGlobMatch glob str = case mspec of
+      Just spec -> checkNameMatches spec glob str
+      Nothing -> if matchGlob glob str then Just (GlobMatch ()) else Nothing
+
+    go (GlobFile glob) dir = do
+      entries <- getDirectoryContents (root </> dir)
+      catMaybes
+        <$> mapM
+          ( \s -> do
+              -- When running a glob from a Cabal package description (i.e.
+              -- when a cabal spec version is passed as an argument), we
+              -- disallow matching a @GlobFile@ against a directory, preferring
+              -- @GlobDir dir GlobDirTrailing@ to specify a directory match.
+              isFile <- maybe (return True) (const $ doesFileExist (root </> dir </> s)) mspec
```

Comment:
This tests if the file exists and is not a directory, but later we might return `GlobMatchesDirectory` ... without checking if the directory exists? Do I read this right?

Reply:
We are sure that the directory is, because we only run this predicate on the results of
```
entries <- getDirectoryContents (root </> dir)
```
which are definitely things that exist.